core/src/impl/Kokkos_FunctorAnalysis.hpp

```diff
@@ -625,29 +637,41 @@ struct FunctorAnalysis {
     }
   };
 
+  template <class F>
+  struct DeduceJoinNoTag<F, std::enable_if_t<(is_reducer<F>::value ||
+                                              (!is_reducer<F>::value &&
+                                               std::is_void<Tag>::value)) &&
+                                             detected_join_no_tag<F>::value>>
+      : public has_join_no_tag_function<F> {
+    enum : bool { value = true };
+  };
+
   template <class F>
   struct DeduceJoinNoTag<
       F,
       std::enable_if_t<(is_reducer<F>::value ||
                         (!is_reducer<F>::value && std::is_void<Tag>::value)) &&
-                       (detected_join_no_tag<F>::value ||
+                       (!detected_join_no_tag<F>::value &&
                         detected_volatile_join_no_tag<F>::value)>>
-      : public std::conditional_t<detected_join_no_tag<F>::value,
-                                  has_join_no_tag_function<F>,
-                                  has_volatile_join_no_tag_function<F>> {
+      : public has_volatile_join_no_tag_function<F> {
     enum : bool { value = true };
   };
 
   template <class F = Functor, typename = void>
   struct DeduceJoin : public DeduceJoinNoTag<F> {};
 
+  template <class F>
+  struct DeduceJoin<
+      F, std::enable_if_t<!is_reducer<F>::value && detected_join_tag<F>::value>>
+      : public has_join_tag_function<F> {
+    enum : bool { value = true };
+  };
+
   template <class F>
   struct DeduceJoin<F, std::enable_if_t<!is_reducer<F>::value &&
-                                        (detected_join_tag<F>::value ||
+                                        (!detected_join_tag<F>::value &&
                                          detected_volatile_join_tag<F>::value)>>
-      : public std::conditional_t<detected_join_tag<F>::value,
-                                  has_join_tag_function<F>,
-                                  has_volatile_join_tag_function<F>> {
+      : public has_volatile_join_tag_function<F> {
```

Comment:
Indeed, this works to fail compilation if this specialization would be selected rather than just warning
```
diff --git a/core/src/impl/Kokkos_FunctorAnalysis.hpp b/core/src/impl/Kokkos_FunctorAnalysis.hpp
index 39891c683..6aef679d2 100644
--- a/core/src/impl/Kokkos_FunctorAnalysis.hpp
+++ b/core/src/impl/Kokkos_FunctorAnalysis.hpp
@@ -655,6 +655,7 @@ struct FunctorAnalysis {
                         detected_volatile_join_no_tag<F>::value)>>
       : public has_volatile_join_no_tag_function<F> {
     enum : bool { value = true };
+    static_assert(!sizeof(F*), "Reducer with a join() operator taking volatile-qualified parameters is no longer supported");
   };
 
   template <class F = Functor, typename = void>
```

The weird `!sizeof(F*)` to express `false` is to make the expression dependent on the type of `F`, and thus only applicable at instantiation and not first-pass semantic analysis of the template definition.

Reply:
https://devblogs.microsoft.com/oldnewthing/20200311-00/?p=103553